distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Column_Naming_Helper.enso

```diff
@@ -110,12 +110,59 @@ type Column_Naming_Helper
        to display every part.
     concat : Vector Text -> Boolean -> Text
     concat self texts add_spaces=True =
-        # TODO truncate parts smarter way
         separator = if add_spaces then " " else ""
         joined = texts.join separator
         case self.naming_properties.size_limit of
             Nothing  -> joined
-            max_size -> self.naming_properties.truncate joined max_size
+            max_size ->
+                base_size = self.naming_properties.encoded_size joined
+                # If it still fits the limit, we are ok.
+                if base_size <= max_size then joined else
+                    # If not, we need to truncate.
+                    separator_size = case add_spaces of
+                        False -> 0
+                        True -> self.naming_properties.encoded_size separator
+
+                    # Estimate the space left for each part to be concatenated.
+                    remaining_space = max_size - (separator_size * (texts.length - 1))
+                    min_bytes_per_part = 5
+
+                    ## If there are so many parts, each of them has very little space,
+                       we cannot fit all of them - so we just keep the first few and the last one.
+                    parts_to_include = case remaining_space <= min_bytes_per_part * texts.length of
+                        # If there is too little space, we will just pick a few parts:
+                        True ->
+                            mid = Math.min (texts.length-1) (remaining_space.div min_bytes_per_part)
+                            texts.take (First 1+mid) + texts.take (Last 1)
+                        False -> texts
+                    new_remaining_space = max_size - (separator_size * (parts_to_include.length - 1))
+                    initial_size_per_part = new_remaining_space.div parts_to_include.length
+
+                    # Now we compute how many parts will fit the initial_size_per_part without truncating - these shorter parts fit as a whole.
+                    part_sizes = parts_to_include.map self.naming_properties.encoded_size
+                    needs_truncating = part_sizes.map (size-> size > initial_size_per_part)
+                    non_truncated_parts_size = part_sizes.zip needs_truncating (size-> is_truncated-> if is_truncated then 0 else size) . fold 0 (+)
+                    truncated_parts_count = needs_truncating.fold 0 acc-> is_truncated->
+                        if is_truncated then acc+1 else acc
+
+                    ## Now having accounted for the parts that do not need truncation,
+                       we distribute the remaining space among the ones that do.
```

Comment:
This is all of course based on estimates and may very well still generate texts that do not fully use the space that is available for the name. 

But that is not the point of this algorithm - the idea is to exploit the available space reasonably well to generate a name that +- shows what is the performed operation even if argument column names are large. We don't need it to be perfect.